Ipa interface overview (#1)
KevinUTAT/oneAPI-samples_wenkaixu_fork: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/IPAuthoring_interfaces_overview/csr-pipes/Readme.md

> # CSR-Pipe

Reviewer: ```suggestion
# CSR Data
```

> This example demonstrate simple vector addition kernel that write to the Control Status Register (CSR) using host pipe interface (CSR-Pipe)

Reviewer: ```suggestion
This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to return output into the CSR using the `protocol_avalon_mm` pipe protocol.
```

## Invocation Interface

> By default, an un-decorated oneAPI kernel will have all its control signals and arguments passed through an MM-Agent interface.

Reviewer: ```suggestion
By default, an un-decorated oneAPI kernel will have all its control signals and arguments mapped into the IP component's control/status register (CSR).
```

Also, please include a figure from here (you may need to customize... let me know if you can't find one in here
https://intel-my.sharepoint.com/personal/paul_white_intel_com/Documents/Documentation/HLS%20Docs/kernel-top-levels.vsdx?web=1

## Data Interface - CSR Pipe

> In this design, input `InputPipeA` and `InputPipeB` are implemented as regular host pipes. However, the output `OutputPipeC` is implemented to write to the kernel CSR. 

Reviewer: ```suggestion
In this design, input `InputPipeA` and `InputPipeB` are implemented as host pipes, as in [Streaming Data](host-pipes). However, the output `OutputPipeC` is implemented to write a single vector sum to the IP component's CSR.
```

> To specify a host pipe to use CSR, set it's protocol as `AVALON_MM` or `AVALON_MM_USES_READY` (from `sycl::ext::intel::prototype::internal` namespace). See how to [specify pipe protocol](../host-pipes/Readme.md#Host-Pipe-Properties).

Reviewer: ```suggestion
To configure a host pipe to map to the CSR, set its protocol as `AVALON_MM` or `AVALON_MM_USES_READY` (from `sycl::ext::intel::prototype::internal` namespace). See how to [specify pipe protocol](../host-pipes/Readme.md#Host-Pipe-Properties).
```

Alternatively, you can use the follwing property shorthands:
- `sycl::ext::intel::experimental::protocol_avalon_mm_uses_ready`
- `sycl::ext::intel::experimental::protocol_avalon_mm`

> For more detail on how to use host pipe in general, refer to the [host-pipes](../host-pipes/) part of this code sample.

Reviewer: ```suggestion
For more detail on how to use host pipes in general, refer to the [Streaming Data](../host-pipes/) part of this code sample.
```

### On Linux*

1. Change to the sample directory.
2. Configure the build system for the Agilex™ 7 device family, which is the default.

   ```
   mkdir build
   cd build
   cmake ..
   ```

   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ``` 
   >
   > Alternatively, you can target an explicit FPGA board variant and BSP by using the following command: 
   >  ```
   >  cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant> -DIS_BSP=1
   >  ``` 
   >
   > You will only be able to run an executable on the FPGA if you specified a BSP.

3. Compile the design. (The provided targets match the recommended development flow.)

   1. Compile for emulation (fast compile time, targets emulated FPGA device).
      ```
      make fpga_emu
      ```

   2. Compile for simulation (fast compile time, targets simulator FPGA device):
      ```
      make fpga_sim
      ```

   3. Generate HTML performance report.
      ```
      make report
      ```
      The report resides at `vector_add.report.prj/reports/report.html`.

>    4. Compile for FPGA hardware (longer compile time, targets FPGA device).
>       ```
>       make fpga
>       ```

Reviewer: ```suggestion
```

not supported

Reviewer 2: @KevinUTAT I still see these "make fpga" and "nmake fpga" (for Window) in all readme. So I marked it as unresolved for now. Please comment.

## Run the Design

### On Linux

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./vector_add.fpga_emu
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./vector_add.fpga_sim
   ```

#### Run on FPGA

1. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   ./vector_add.fpga
   ```

### On Windows

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   vector_add.fpga_emu.exe
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   vector_add.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```
#### Run on FPGA

1. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   vector_add.fpga.exe
   ```

## Example Output

```
add two vectors of size 256
PASSED
```

> 

Reviewer: add a note about how to see the output entry in the CSR hpp files
